### DISCUSSION TITLE
Выполнение HTTP запросов с передачей данных внутрь запроса.

### DISCUSSION THREAD

**Discussion by Qb1cc** [totumonline/totum-ru-issues-and-discussions#78] [Как это сделать?] [answered]
Для большинства sms-шлюзов, требуется выполнять HTTP запросы, как выполнить их в тотуме и передать в него данные из таблицы.

`http://sms.ru/sms/send?api_id****&to=79255070602(получаем из тотум)&msg=Ваш заказ (получаем сюда номер цикла) готов к выдаче.&json=1`

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Это делается функцией [getFromScript](https://docs.totum.online/functions#fn-getFromScript).

Те например нам нужно отправить смс если статус у поля стал ревен 3. В код-действия при изменении поля статуса пишем:

```
a1=: if(condition: $#nfv = 3; then: $sms)

	sms: getFromScript(uri: "http://sms.ru/sms/send"; gets: $get)
		get: rowCreate(field: "api_id" = "*****"; field: "to" = $tel; field: "msg" = str`"Ваш заказ" ++ $number ++ "готов к выдаче"`; field: "json" = 1)
		
			// Если как здесь из поля берется то прям внутрь функции засовываете — не нужно отдельную строку. Отдельная строка если там select итд.
			tel: #telephone
			number: #number

```

Параметры после `?` это `GET` — поэтому в `gets`.

Про `str` [вот здесь!](https://docs.totum.online/using-str-to-combine-text)

Это самый простой подход и у него есть недостаток — вызов стороннего сервера осуществляется в момент смены статуса и если этот сторонний сервер тупит, то тупить будет и база Тотума.

Поэтому прогрессивный способ это делать — это при смене статуса добавляете строку с необходимыми параметрами в отдельную таблицу и там ставите чекбоксы [кроном](https://docs.totum.online/crons) раз в минуту и этот код действия пишете в чекбокс.

Если вам не нужно гарантировать 100% доставку сообщения до стороннего сервера, то можно в код выше добавить `ssh: true` в  [getFromScript](https://docs.totum.online/functions#fn-getFromScript) — тогда этот вызов будет отделен в отдельный процесс на сервере и если там что-то пойдет не так, то он просто умрет и это никак не повлияет на пользователя Тотум тк в этом случае Тотум не ждет его полного выполнения. Но в этом случае Тотум не получает возврата результата работы этого скрипта — использовать только для односторонних неответственных хуков!

**Comment by Qb1cc** [comment]
Если понадобится полная инструкция, рабочая, как привязать к этому всему свой модем с симкой, Пишите, я подготовлю полную инструкцию по настройке шлюза(всё на одном хосте лежит и без интранетов работает) с USB модемом. По итогу вы получите почти халявные смски оповещения ( но если вы хотите именные смс,  то дешевле использовать представленные шлюзы в интернетах и необходимость настраивать собственные шлюзы отпадает )
